machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,131 @@
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuild describes a build process managed by the MCO
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuild struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machine os build
+	// +kubebuilder:validation:Required
+	Spec MachineOSBuildSpec `json:"spec"`
+
+	// status describes the lst observed state of this machine os build
+	// +optional
+	Status MachineOSBuildStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineOSObjectReference `json:"machineConfigPool"`
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSObjectReference `json:"renderedMachineOSImage"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig string `json:"currentConfig"`
+	// desiredConfig is the desired config we want to build an image for. If currentConfig and desiredConfig are not the same, we need to build an image.
+	// +kubebuilder:validation:Required
+	DesiredConfig string `json:"desiredConfig"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// buildStart describes when this build began
+	// +kubebuilder:validation:Required
+	BuildStartTime metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended
+	// +optional
+	BuildEndTime metav1.Time `json:"buildEnd"`
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Built
```

Comment:
I wonder if you need the `Build` prefix on all of these, given the context of the object, WDYT?

Suggestion, maybe instead of `Built` you could use `Ready` which is similar to other APIs.

I'd be interested to learn more about the other condition types, I wonder if some may be better consolidated, interrupted and restarted perhaps are failure conditions of Building?

Reply:
I agree. I don't think the `Build` prefix is needed here.